src/Microsoft.AspNetCore.OData/Batch/ODataBatchReaderExtensions.cs

```diff
@@ -236,5 +268,110 @@ private static HttpContext CreateHttpContext(HttpContext originalContext)
+        /// request is retained and the one from the batch is discarded.
+        /// </summary>
+        /// <param name="individualPreferences">The value of the Prefer header from the individual request inside the batch</param>
+        /// <param name="batchPreferences">The value of the Prefer header from the overall batch request</param>
+        /// <returns>Value containing the combined preferences</returns>
+        private static string MergeIndividualAndBatchPreferences(string individualPreferences, string batchPreferences)
+        {
+            if (string.IsNullOrEmpty(individualPreferences))
+            {
+                return batchPreferences;
+            }
+
+            if (string.IsNullOrEmpty(batchPreferences))
+            {
+                return individualPreferences;
+            }
+            // get the name of each preference to avoid adding duplicates from batch
+            IEnumerable<string> individualList = individualPreferences.SplitPreferences().Select(pref => pref.Trim());
+            HashSet<string> individualPreferenceNames = new HashSet<string>(individualList.Select(pref => pref.Split('=').FirstOrDefault()));
+
+            
+            IEnumerable<string> filteredBatchList = batchPreferences.SplitPreferences().Select(pref => pref.Trim())
+                // do not add duplicate preferences from batch
+                .Where(pref => !individualPreferenceNames.Contains(pref.Split('=').FirstOrDefault()));
+            string filteredBatchPreferences = string.Join(", ", filteredBatchList);
+
+            return string.Join(", ", individualPreferences, filteredBatchPreferences);
+        }
+
+        /// <summary>
+        /// Splits the value of a Prefer header into separate preferences
+        /// e.g. a value like 'a, b=c, foo="bar,baz"' will return an IEnumerable with
+        /// - a
+        /// - b=c
+        /// - foo="bar,baz"
+        /// </summary>
+        /// <param name="preferences"></param>
+        /// <returns></returns>
+        private static IEnumerable<string> SplitPreferences(this string preferences)
+        {
+            StringBuilder currentPreference = new StringBuilder();
+            HashSet<string> addedPreferences = new HashSet<string>();
+            bool insideQuotedValue = false;
+            char prevChar = '\0';
+            foreach (char c in preferences)
+            {
+                if (c == '"')
+                {
+                    if (!insideQuotedValue)
+                    {
+                        // we are starting a double-quoted value
+                        insideQuotedValue = true;
+                    }
+                    else
+                    {
+                        // this could be the end of a quoted value, or it could be an escaped quote
+                        insideQuotedValue = prevChar == '\\';
+                    }
+                }
+
+                if (c == ',' && !insideQuotedValue)
```

Comment:
make it an else if... no need to check if we've already determined above that it is a quote.